### PR TITLE
WIP: Add TimeDecoding object to time_gen.py

### PR DESCRIPTION
**Pull request by MechCoder**
Add TimeDecoding to predict the event_id for all epochs for a given time.

TimeDecoding().fit()
Fits a series of estimators (n_folds) across every time point

TimeDecoding().predict()
Predict the event_ids for every epoch.
In "mean-prediction" mode, the predictions are averaged across all folds, while in the cross-validation mode, the predictions are those that are made for each test set. Same thing goes for scores as well.

Plot plots the cross-validation score against time graph.

### CODE REVIEW

**Comment by MechCoder** [comment]
@agramfort I've just added the fit method for now. If I'm on the right track.. I can add predict, score etc..

**Comment by agramfort** [comment]
yes you are on the right track. Can you see how it simplifies the corresponding example?

a simple example is a good sign of a good API

**Comment by MechCoder** [comment]
I'm really not sure if that question is rhetoric or not.. It should be now similar to the GAT example.

**Comment by MechCoder** [comment]
@agramfort I've added support for score, predict, plot and a few minor tests as well. Would be great if you could give a first pass.

cc: @dengemann @Eric89GXL and other core devs..

**Comment by mainakjas** [comment]
Does this simplify the `plot_decoding_sensors.py` example? Any other example it simplifies?

File: mne/decoding/time_gen.py [hunk 1]
```diff
@@ -914,3 +953,358 @@ def time_generalization(epochs_list, clf=None, cv=5, scoring="roc_auc",
...
+        if hasattr(self, 'scores_'):
+            s += "scored"
+            if callable(self.scorer_):
+                s += " (%s)" % (self.scorer_.__name__)
+        else:
+            s += "no score"
+
+        return "<TimeDecoding | %s>" % s
+
+    def fit(self, epochs, y=None, picks=None):
```
Comment (mainakjas) [inline mne/decoding/time_gen.py]: why do a `picks` here? The user can do their picks and then feed it to the classifier, no?

File: mne/decoding/time_gen.py [hunk 2]
```diff
@@ -914,3 +953,358 @@ def time_generalization(epochs_list, clf=None, cv=5, scoring="roc_auc",
...
+    Cross-validate a series of estimators across every time point
+    to decode the event_id of each epoch.
+
+    Parameters
+    ----------
+    cv : int | object
+        If an integer is passed, it is the number of folds.
+        Specific cross-validation objects can be passed, see
+        sklearn.cross_validation module for the list of possible objects.
+        Defaults to 5.
```
Comment (mainakjas) [inline mne/decoding/time_gen.py]: you mean `KFold` cross-validation with `K=5`?

File: mne/decoding/time_gen.py [hunk 3]
```diff
@@ -914,3 +953,358 @@ def time_generalization(epochs_list, clf=None, cv=5, scoring="roc_auc",
                       for train, test in cv)
     scores = np.mean(scores, axis=0)
     return scores
+
+
+class TimeDecoding(_Decoder):
+    """
+    Cross-validate a series of estimators across every time point
+    to decode the event_id of each epoch.
```
Comment (mainakjas) [inline mne/decoding/time_gen.py]: Can you give me a background as to what this class does as opposed to the GAT class?

Comment (agramfort) [inline mne/decoding/time_gen.py]: it's doing less hence is simpler / faster but tries to reuse as much code
as possible the GAT utilities.

File: mne/viz/decoding.py [hunk 4]
```diff
@@ -152,3 +152,79 @@ def plot_gat_diagonal(gat, title=None, xmin=None, xmax=None, ymin=0., ymax=1.,
     if show is True:
         plt.show()
     return fig if ax is None else ax.get_figure()
+
+
+def plot_time_decoding_scores(tdec, title=None, xmin=None, xmax=None,
```
Comment (mainakjas) [inline mne/viz/decoding.py]: I think this should go to the `viz` module

File: mne/viz/decoding.py [hunk 5]
```diff
@@ -152,3 +152,79 @@ def plot_gat_diagonal(gat, title=None, xmin=None, xmax=None, ymin=0., ymax=1.,
...
+
+    if not hasattr(tdec, 'scores_'):
+        raise RuntimeError('Please score your data before trying to plot '
+                           'scores')
+    import matplotlib.pyplot as plt
+
+    if ax is None:
+        fig, ax = plt.subplots(1, 1)
+
+    times = tdec.times_ * 10**3
```
Comment (mainakjas) [inline mne/viz/decoding.py]: woops, what's this?

File: mne/decoding/time_gen.py (hunk 3 above)
Comment (mainakjas) [inline mne/decoding/time_gen.py]: less = what?

File: mne/decoding/time_gen.py [hunk 6]
```diff
@@ -914,3 +953,358 @@ def time_generalization(epochs_list, clf=None, cv=5, scoring="roc_auc",
...
+    predict_mode : {'cross-validation', 'mean-prediction'}
+        Indicates how predictions are achieved with regards to the cross-
+        validation procedure:
+
+            ``cross-validation`` : estimates a single prediction per sample
+                based on the unique independent classifier fitted in the
+                cross-validation.
+            ``mean-prediction`` : estimates k predictions per sample, based on
+                each of the k-fold cross-validation classifiers, and average
+                these predictions into a single estimate per sample.
```
Comment (mainakjas) [inline mne/decoding/time_gen.py]: okay, so this is supposed to work only for probabilistic estimates, but not for categorical estimates? Is the average user going to need this option?

File: mne/decoding/time_gen.py [hunk 7]
```diff
@@ -914,3 +953,358 @@ def time_generalization(epochs_list, clf=None, cv=5, scoring="roc_auc",
...
+        ----------
+        epochs : instance of Epochs
+            The epochs. Can be similar to fitted epochs or not. See independent
+            parameter.
+
+        picks : np.ndarray (n_selected_chans,) | None
+            Channels to be included.
+
+        test_times : None
+            Not to be used.
```
Comment (mainakjas) [inline mne/decoding/time_gen.py]: why have this option then?

File: mne/decoding/time_gen.py [hunk 8]
```diff
@@ -914,3 +953,358 @@ def time_generalization(epochs_list, clf=None, cv=5, scoring="roc_auc",
...
+            Class labels for samples in X.
+        """
+        from scipy import stats
+        n_jobs = self.n_jobs
+
+        if not hasattr(self, 'estimators_'):
+            raise ValueError("Estimators have to be fit before predicting.")
+
+        if picks is None:
+            picks = pick_types(epochs.info, meg=True, eeg=True,
```
Comment (mainakjas) [inline mne/decoding/time_gen.py]: we should probably have some sort of check to see if the `info` object is the same for training and testing

**Comment by mainakjas** [comment]
Travis fails. That's it from me for now. If you update an example, I can give it a try. Thanks for the contribution!

**Comment by kingjr** [comment]
Why not just making a class from GAT? It's already doing this no?

```
class TimeDecoding(GeneralizationAcrossTime):

    def predict(self, X, test_times='diagonal', **kwargs):
        super(TimeDecoding, self).predict(X, test_times, **kwargs)

    def plot(self, **kwargs):
        super(TimeDecoding, self).plot_diagonal(**kwargs)
```


full example:

```
"""
======================================
Decoding sensor space data Across Time
======================================

This example runs the analysis computed in:

Jean-Remi King, Alexandre Gramfort, Aaron Schurger, Lionel Naccache
and Stanislas Dehaene, "Two distinct dynamic modes subtend the detection of
unexpected sounds", PLOS ONE, 2013,
http://www.ncbi.nlm.nih.gov/pubmed/24475052

The idea is to decode the MEG pattern at each time point.
"""
# Authors: Jean-Remi King <jeanremi.king@gmail.com>
#
# License: BSD (3-clause)

import mne
from mne.datasets import spm_face
from mne.decoding import GeneralizationAcrossTime

print(__doc__)

# Preprocess data
data_path = spm_face.data_path()
# Load and filter data, set up epochs
raw_fname = data_path + '/MEG/spm/SPM_CTF_MEG_example_faces%d_3D_raw.fif'
events_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw-eve.fif'

raw = mne.io.Raw(raw_fname % 1, preload=True)  # Take first run

picks = mne.pick_types(raw.info, meg=True, exclude='bads')
raw.filter(1, 45, method='iir')

events = mne.find_events(raw, stim_channel='UPPT001')
event_id = {"faces": 1, "scrambled": 2}
tmin, tmax = -0.1, 0.5

decim = 4  # decimate to make the example faster to run
epochs = mne.Epochs(raw, events, event_id, tmin, tmax, proj=True,
                    picks=picks, baseline=None, preload=True,
                    reject=dict(mag=1.5e-12), decim=decim, verbose=False)

# Define decoder. The decision_function is employed to use AUC for scoring


class TimeDecoding(GeneralizationAcrossTime):

    def predict(self, X, test_times='diagonal', **kwargs):
        super(TimeDecoding, self).predict(X, test_times, **kwargs)

    def plot(self, **kwargs):
        super(TimeDecoding, self).plot_diagonal(**kwargs)


tg = TimeDecoding(predict_mode='cross-validation', n_jobs=-1)
# fit and score
tg.fit(epochs)
tg.score(epochs)
tg.plot_diagonal()
```

File: mne/decoding/time_gen.py (hunk 6 above)
Comment (dengemann) [inline mne/decoding/time_gen.py]: I think we just trimmed down the prediction API with @kingjr -- let's make sure to be consistent.

File: mne/decoding/time_gen.py [hunk 9]
```diff
@@ -914,3 +953,358 @@ def time_generalization(epochs_list, clf=None, cv=5, scoring="roc_auc",
...
+                    if self.predict_type == "predict":
+                        predicted[test, :] = estimator.predict(
+                            X[test, :])[:, None]
+                    elif self.predict_type == "predict_proba":
+                        predicted[test, :] = estimator.predict_proba(
+                            X[test, :])
+                    else:
+                        predicted[test, :] = estimator.decision_function(
+                            X[test, :])[:, None]
+                return predicted
```
Comment (dengemann) [inline mne/decoding/time_gen.py]: I think we meanwhile agreed on not adding nested functions unless we really need to.

**Comment by dengemann** [comment]
One question. What's the difference between this and the GAT if you leave away the off-diagonal estimations?
If there is not any difference I would prefer a wrapper that calls GAT with parameters that avoid off-diagnoal testing.

@kingjr it should not be too difficult to call GAT in such away, maybe if train times and test times are scalr 1 or so?

**Comment by dengemann** [comment]
Ideally we could avoid reduplicating code paths. To me time decoding should do the same as GAT without the off-diagonal cases. Having fewer code paths is better for testing and everything else.

**Comment by kingjr** [comment]
see https://github.com/mne-tools/mne-python/pull/2066

**Comment by dengemann** [comment]
> see #2066

Oh boy, indeed. It seems we're reduplicating efforts here and adding many unnecessary lines.
If there's something new here please put it in GAT.

**Comment by dengemann** [comment]
changed to WIP.

**Comment by kingjr** [comment]
I just did it now, as I didn't see why we just rewrite stuff. GAT is already doing the job, and in fact already has a test_times='diagonal' option.

**Comment by dengemann** [comment]
Ok @kingjr @MechCoder please coordinate. It seems we can get some new nice interface less expensively.

**Comment by dengemann** [comment]
+100 for having one code path to do all this. I don't really want to think about wheather diagonal GAT and   TimeDecoding do the same or not. 

```python -c "import this" | grep "only"```

**Comment by kingjr** [comment]
Agreed. Let s move the test_times='diagonal' from the GAT to the
DecodingTime class.

I ll do this tomorrow

**Comment by dengemann** [comment]
> Agreed. Let s move the test_times='diagonal' from the GAT to the
DecodingTime class.

**Comment by dengemann** [comment]
> Agreed. Let s move the test_times='diagonal' from the GAT to the
DecodingTime class.

I was trying to say that we should have one code for everything behind the scenes. We should maybe have 2 interfaces. We should not remove the diagonal plotting from GAT. I see this more as a modular thing. Eithe you want only the diagonal or both. In the first case you can use a special object for that. ...

Not sure. Maybe just chaging docs and examples would be enough. Or Having a genral class TimeDecoding that does GAT and | or diagonal. The more I think about it the less I'm convinced that we need 2 objects.

**Comment by agramfort** [comment]
I vote for 2 classes to simplify the options.

@kingjr do you want to take over the PR?

**Comment by agramfort** [comment]
If we can reuse GAT class without doing unnecessary computation I am happy.

**Comment by mainakjas** [comment]
+1 for two separate classes. I don't know how common GAT is, but
time-resolved decoding is pretty common. Also, I'd name the class
TimeResolvedDecoding. TimeDecoding is a bit ambiguous to me.

On Thu, May 7, 2015 at 9:29 AM, Alexandre Gramfort <notifications@github.com

**Comment by agramfort** [comment]
an option that would be useful for TimeDecoding class is prediction on
accumulated data... but maybe later...

**Comment by kingjr** [comment]
I'm also for two classes. If the release hasn't been done yet (?), I think it'd make sense to pass the ```test_times = 'diagonal'``` to the TimeDecoding object: GAT would be the generic class with a strict but simple time constructors, and the TimeDecoding would construct diagonal testing_time.

> Also, I'd name the class TimeResolvedDecoding. TimeDecoding is a bit ambiguous to me.

I Agree. Other possibilites are TimeSearchLignt (to bridge to fMRI), TemporalDecoding (although ambiguous with temporal cortex?).


> an option that would be useful for TimeDecoding class is prediction on accumulated data... but maybe later...

@agramfort Could you be more specific? The test_times param accepts > 1 sample per classifer.

**Comment by agramfort** [comment]
> I'm also for two classes. If the release hasn't been done yet (?), I think it'd make sense to pass the test_times = 'diagonal' to the TimeDecoding object: GAT would be the generic class with a strict but simple time constructors, and the TimeDecoding would construct diagonal testing_time.

generic with simple in same sentence? :)

first why in the first place the GAT class can be used without
generalization? to me it's a bug more than a feature.

> Also, I'd name the class TimeResolvedDecoding. TimeDecoding is a bit ambiguous to me.

no opinion here.

> I Agree. Other possibilites are TimeSearchLignt (to bridge to fMRI), TemporalDecoding (although ambiguous with temporal cortex?).

-1 on search light

> an option that would be useful for TimeDecoding class is prediction on accumulated data... but maybe later...
>
> @agramfort Could you be more specific? The test_times param accepts > 1 sample per classifer.

learn from with data from t0 to t rather than just at t

**Comment by dengemann** [comment]
We should use a private class which does all, the GAT would then expose the
generalization (not allow for purely diag decoding), the TimeDecoding
should only expose the diagnoal decoding.

2015-05-07 14:17 GMT+02:00 Alexandre Gramfort <notifications@github.com>:

**Comment by kingjr** [comment]
> We should use a private class which does all, the GAT would then expose the generalization (not allow for purely diag decoding), the TimeDecoding should only expose the diagnoal decoding.

I agree. This would basically mean:
1. renaming the current GAT class and make it private
2. creating two new public classes: i) TimeResolved ii) GeneralizationAcrossTime 

However, in this scenario, I'm not sure what would be specifically implemented in the GAT public class that wouldn't be in the private one: a check that at least one classifier is tested at some other time? That seems a bit useless?

> learn from with data from t0 to t rather than just at t

Ok. This is already compatible with the current script, but you have to defined ```test_times``` manually. We can add a ```test_times='cumulative'``` constructor to make it easier; or yet create another TimeCumulativeDecoding class?.. Although one could say that cumulative decoding is qualititavely distinct from TimeResolvedDecoding...

**Comment by agramfort** [comment]
> We should use a private class which does all, the GAT would then expose the generalization (not allow for purely diag decoding), the TimeDecoding should only expose the diagnoal decoding.
>
> I agree. This would basically mean:
> 1. renaming the current GAT class and make it private
> 2. creating two new public classes: i) TimeResolved ii) GeneralizationAcrossTime
>
> However, in this scenario, I'm not sure what would be specifically implemented in the GAT public class that wouldn't be in the private one: a check that at least one classifier is tested at some other time? That seems a bit useless?

any chance we can simplify the API of GAT in this case? if not it's
indeed useless.

> learn from with data from t0 to t rather than just at t
>
> Ok. This is already compatible with the current script, but you have to defined test_times manually. We can add a test_times='cumulative' constructor to make it easier; or yet create another TimeCumulativeDecoding class?.. Although one could say that cumulative decoding is qualititavely distinct from TimeResolvedDecoding...

I would go with another class that inherits from a base time decoding
class... but I don't see an urgent matter given the need to release
asap.

**Comment by kingjr** [comment]
> I would go with another class that inherits from a base time decoding class... but I don't see an urgent matter given the need to release asap.

Ok let's keep this for a later. 

> any chance we can simplify the API of GAT in this case? if not it's indeed useless.
I can't think of anything right now. @dengemann Do you have any idea there?
